scripts/conditionalMitigator/conditionalMitigator.ps1

```diff
@@ -0,0 +1,391 @@
+    LogWithTimeStamp -msgStr "Checking if any endpoints have been deleted"
+    $stalePortIdList = @()
+    foreach ($portId in $g_endpointInfoMap.Keys) {
+        $portIdPresent = $false
+        foreach ($vfpPort in $vfpPortList) {
+            if ($vfpPort.Id -eq $portId) {
+                $portIdPresent = $true
+                break
+            }
+        }
+
+        if ($portIdPresent -eq $false) {
+            $stalePortIdList += @($portId)
+        }
+    }
+    $priorSize = $g_endpointInfoMap.count
+    foreach ($portId in $stalePortIdList) {
+        $g_endpointInfoMap.Remove($portId)
+    }
+
+    $endpointsDeleted = $priorSize - $g_endpointInfoMap.count
+    LogWithTimeStamp -msgStr ("old endpoints deleted from g_endpointInfoMap: {0}" -f $endpointsDeleted)
+    ##
+}
+
+function RulesAreMissing() {
+    ## Check pod port rules.
+    foreach ($portId in $g_endpointInfoMap.Keys)
+    {
+        $isPodPort = PortIsPodPort -vfpPortInfo $g_currentVfpPortMap[$portId]
+        if ($isPodPort -eq $false) {
+            # this could be external port or host vNIC. Ignore.
+            continue
+        }
+
+        $current_time = get-date
+        if ($g_endpointInfoMap.ruleCheckCount -gt 0) {
+            $timeSinceLastCheck = $current_time - $g_endpointInfoMap[$portId].lastRuleCheckTime
+            if ($timeSinceLastCheck.TotalSeconds -lt $RuleCheckIntervalSecs) {
+                # check again later
+                continue
+            }
+        } else {
+            $timeSinceLastCheck = $current_time - $g_scriptStartTime
+        }
+
+        $rulesPresent = CheckForRulesOnVfpPort -portId $portId -rulesToCheck $g_podRuleCheckList
+        $g_endpointInfoMap[$portId].ruleCheckCount += 1
+
+        if ($rulesPresent -eq $false) {
+            # We reach here when a port does not have the necessary rules for more than RuleCheckIntervalMins.
+            # Mitigation action must be taken.
+            LogWithTimeStamp -msgStr ("Rules missing on VFP port with ID {0} since atleast last {1:N2} minutes" -f $portId,$timeSinceLastCheck.TotalMinutes)
+            return $true
+        }
+
+        $g_endpointInfoMap[$portId].lastRuleCheckTime = $current_time
+        # This port has the necessary rules.
+    }
+    ## Pod port rule check done.
+
+    return $false
+}
+
+
+function ScriptSetup()
+{
+    foreach ($rule in $PodPortRulesToCheck) {
+        $ruleCheckInfo = [RuleCheckInfo]::New($rule.ruleRegex, $rule.layerName, $rule.groupName)
+        $g_podRuleCheckList.Add($ruleCheckInfo)
+    }
+    LogWithTimeStamp -msgStr ("Number of pod port rules to check: {0}" -f $g_podRuleCheckList.count)
+}
+
+
+function CheckIfMitigationRequired()
+{
+    NoteCurrentVfpPorts
+
+    $rulesMissing = RulesAreMissing
+    return $rulesMissing
+}
+
+
+function collectLogsBeforeMitigation(
+    [string]$LogsPath
+)
+{
+    if ($CollectWindowsLogs -eq $true) {
+        # create log path if not yet created.
+        mkdir -Force $LogsPath
```

Comment:
If $CollectWindowsLogs is $true and issue happens very frequently, there are chances multiple log files will be created. This will consume lot of customer's space.

Reply:
Yes, we will need to add a rotation logic for the same. Will add that.